File: python/031-040/Combination Sum.py
#!/usr/bin/env python
# coding=utf-8

class Solution:
    # @param {integer[]} candidates
    # @param {integer} target
    # @return {integer[][]}
    def combinationSum(self, candidates, target):
        candidates.sort()
        result = []
        tmp = []
        self.help(tmp, candidates, 0, target, result)
        return result
    
    def help(self, now, candidates, index, target, result):
        for i in range(index, len(candidates)):
            if candidates[i] < target:
                now.append(candidates[i]);
                self.help(now, candidates, i, target-candidates[i], result)
                del now[-1]
            elif candidates[i] == target:
                now.append(candidates[i])
                temp = now[:]
                result.append(temp)
                del now[-1]

File: cpp/031-040/Combination Sum.cpp
class Solution {
    
    void help(vector<int>& now, vector<int>&candidates, int index, int target, vector<vector<int>>& results)
    {
        for (int i = index;i < candidates.size();i++)
        {
            if (candidates[i] < target)
            {
                now.push_back(candidates[i]);
                help(now, candidates, i, target-candidates[i], results);
                now.pop_back();
            } else if (candidates[i] == target) {
                now.push_back(candidates[i]);
                results.push_back(now);
                now.pop_back();
            }
        }
    }
    
public:
    vector<vector<int> > combinationSum(vector<int> &candidates, int target) {
        vector<vector<int>> results;
        vector<int>tmp;
        help(tmp, candidates, 0, target, results);
        return results;
    }
    
};
